src/systems/character-controller-system.js

```diff
@@ -250,25 +250,37 @@ export class CharacterControllerSystem {
       rotateInPlaceAroundWorldUp(this.avatarPOV.object3D.matrixWorld, this.dXZ, snapRotatedPOV);
 
       const playerScale = v.setFromMatrixColumn(this.avatarPOV.object3D.matrixWorld, 1).length();
-      calculateDisplacementToDesiredPOV(
-        snapRotatedPOV,
-        this.fly,
-        this.relativeMotion.multiplyScalar(
-          ((userinput.get(paths.actions.boost) ? 2 : 1) * BASE_SPEED * Math.sqrt(playerScale) * dt) / 1000
-        ),
-        displacementToDesiredPOV
-      );
+      if (this.relativeMotion.lengthSq() > 0.000001) {
+        calculateDisplacementToDesiredPOV(
+          snapRotatedPOV,
+          this.fly,
+          this.relativeMotion.multiplyScalar(
+            ((userinput.get(paths.actions.boost) ? 2 : 1) * BASE_SPEED * Math.sqrt(playerScale) * dt) / 1000
+          ),
+          displacementToDesiredPOV
+        );
+      } else {
+        displacementToDesiredPOV.set(0, 0, 0);
```

Comment:
Rather than setting this to 0 and then checking the length below, why not just move all the work into the if, and if you don't have a large enough relative motion, just do nothing.

Reply:
we still need to snap to nav mesh if you changed in/out of fly mode, or the land flag is set from a previous frame